### DISCUSSION TITLE
osu!taiko/mania multiple keys on a keybind (stable)

### DISCUSSION THREAD

**Discussion by Mazzuli500** [ppy/osu#23676] [Q&A]
surprised to find literally no one talk about this

from thread: https://github.com/ppy/osu/discussions/14597
"We've historically allowed players to bind more keys via external programs as long as the extra keys don't affect the already-pressed keys, such as releasing them."
Two similar questions:

1. 
about taiko mouse keybinds:
Does this mean you can also use external software to add a left/right mouse button key on your keyboard in taiko (stable)?
The reason why I'm asking this is because it has a much bigger impact on taiko than standard, because in osu!taiko (stable) the mouse buttons basically work as a third don/kat button
("-" in keybinds separates hands to make it easier to read)
If it is allowed: players with dd-kk binds will feel the biggest impact since they can simply add another button, making it 
ddd-kkk
this would make certain patterns way easier than with any playstyle with only 4 keys, example:
![image](https://github.com/ppy/osu/assets/67278477/0f10a993-d5d2-4dcc-99dc-51bb79e8626d)

Other playstyles (such as kd-dk) will get two more buttons to play with, but I think it would be a headache to learn any other keybind combination than ddd-kkk (kdd-dkk for example)
So effectively dd-kk players get buffed a lot and everyone else gets buffed but only by a little.
Yes, osu!taiko (lazer) has three keybinds for each color but two of the keybinds are for one hand and they're handled similarly to osu!standard mouse inputs.

2. 
Is is allowed to use external software to have more than one button on your keyboard do the same thing on taiko/mania? 
I figured I should ask this specifically about these modes because contrary to stable, you can use two hands to hit notes without any problem on taiko and mania.
first example that comes to mind for taiko: 
kdkd-dkdk, so that one hand has all unique keybinds. Releasing your left hand keys before pressing the right hand keys is relatively easy, so on certain patterns speed can be effectively doubled.
mania 4k example:
with the keybinds (from left to right) d,f,h,j on your keyboard, you can add secondary keys on (from left to right) k,l,a,s
This gives an obvious advantage on one-hand trill patterns:
![image](https://github.com/ppy/osu/assets/67278477/5ef9e953-cf8d-499c-b7c8-6528d18ff79f)
 (instead of alternating left middle/index fingers, you can do left middle, left index, right ring, right pinky, repeat)
and vibro patterns:
![image](https://github.com/ppy/osu/assets/67278477/7ab6b60d-46f9-48f0-b96a-e9ffb723090d)
 (instead of having to vibrating your hands, you can alternate using your left and right hand)

using more than one finger on one keybind is already used by basically every 1 key mania player (yes they exist) so that is not an excuse
As for people who would actually use these, I would. I just don't want to learn a playstyle that might get banned later.

**Comment by Mazzuli500** [comment]
If both of these are allowed, well it has an interesting effect on taiko as a whole.
Instead of people being either kddk or ddkk players, you'll see players with kdkddddkdk, kkkddddkkk etc. Keybinds
That is interesting in itself but also very problematic

**Comment by aryoadhi** [comment]
Please no. This is probably an idea for multi-binding cheating. The only multiple keybind that is allowed on mania is special key binding, due to it effectively emulates scratch up/down motion on special keys (5K+1, 7K+1, 10K+2, 14K+2).

I don't see the reason to add this feature, especially if this greatly benefits players who uses this control scheme. If the game is already designed that way, adding option for multi-key binding is the least thing me as a player want to see.

**Reply by Mazzuli500** [reply]
But is multi-binding really not allowed? How I see it is that multi-binds are essentially the same as hitting one key with more than one finger, because in both scenarios you have to account for releasing the note before pressing it again. 
So if one is allowed, the other one should also be allowed. afaik there isn't any rule against pressing a single keybind with more than one finger?
Maybe just disallow people from pressing the same input with more than one finger? that would basically negate the 2nd question entirely. But maybe allow it for 1-3k because they are not rankable anyway and they rely on it?

**Reply by aryoadhi** [reply]
Multi-binding and button mashing is entirely different playstyle. For example, consider this case:

A 7K map contains a series of jacks in middle/thumb button. The mapper wants the players to play it intended way, by swapping thumb presses on the spacebar. It poses its own challenge as players need to take account of how long the key will depress so they can register new key press rather than extending the hold time. This technique is common in some arcade rhythm games (DDR with foot switch, SDVX with single lane jacks, IIDX with scratch jacks).

With multi binding, this is not an issue since players can theoretically mash two different buttons, in which the physical limitation is greatly reduced. This clearly gives advantage to players who multibind over those who are not.

However, I don't see this is beneficial to the best as current osu!stable system of multibinding (as seen on N+1 keycounts) still plays so bad (as mentioned on https://github.com/ppy/osu/discussions/14597#discussioncomment-1263308), and lazer still doesn't have proper special key support at all. 1-3K rank support still not yet considered as we don't have proper rules and set for such. Maybe we can implement these for lower keycount, but that won't be possible until this behavior altered on mania (pressing new button will immediately release-and-press the bound column).

For taiko, this is extremely beneficial since taiko does not have hold notes on don/kat, only rolls and spinners. Maps with long d/k trills can be cheesed easily.

**Reply by Mazzuli500** [reply]
If it is not allowed, what else is not allowed then?
So the thing that differentiates these playstyles is the fact that the key is physically pressed down on your keyboard?
Hypothetically, what if you had a custom keyboard which works the same as tapping on a table? (the keys don't physically press down)
Then it REALLY would be the same regardless if you alterate your fingers on one key or two keys (which have the same output) since the physical part plays the same in that case.

**Reply by Mazzuli500** [reply]
just to add the thought process of why I think mashing one button or two is basically the same thing:
Imagine a keyboard with really long keys:
![image](https://github.com/ppy/osu/assets/67278477/7fa91767-7a37-41fa-939b-7dd8e6ebda6e)
then you could comfortably put both of your hands on all of the keys, right? And it's the same as with a normal keyboard, just made a bit more comfortable (and I don't see this being bannable)
Now, you can make this more ergonomic by splitting the keys so that when the other part is pressed down physically the same way:
![image](https://github.com/ppy/osu/assets/67278477/0d73729d-84b9-461e-b44e-f3b0f90c9125)

Also, what about split keyboards with a split space bar? That is technically multi-binding if you use both thumbs, is that also disallowed?
![image](https://github.com/ppy/osu/assets/67278477/58bc48ef-37dd-4ce2-9134-25695607878f)
